src/clj/schema/experimental/generators.clj

```diff
@@ -26,56 +31,156 @@
-     (= x s/Keyword) generators/keyword
-     :else (generators/return x))))
+(extend-protocol CompositeGenerator
+  schema.spec.variant.VariantSpec
+  (composite-generator [s params]
+    (generators/such-that
+     #(not ((:pre s) %))
+     (generators/one-of
+      (for [o (macros/safe-get s :options)]
+        (sub-generator o params)))))
 
-(extend-type schema.core.EqSchema
-  LeafGeneratable
-  (leaf-generator [x]
-    (generators/return (:v x))))
+  ;; TODO: this does not currently capture proper semantics of maps with
+  ;; both specific keys and key schemas that can override them.
+  schema.spec.collection.CollectionSpec
+  (composite-generator [s params]
+    (generators/such-that
+     #(not ((:pre s) %))
+     (generators/fmap
+      (comp (:constructor s) (partial apply concat))
+      (apply
+       generators/tuple
+       (for [{:keys [cardinality] :as e} (:elements s)]
+         (let [g (sub-generator e params)]
+           (case cardinality
+             :exactly-one (generators/fmap vector g)
+             :at-most-one (generators/one-of [(generators/return nil) (generators/fmap vector g)])
+             :zero-or-more (generators/vector g))))))))
 
-(defprotocol Generator
-  (generator [s params]))
+  schema.spec.leaf.LeafSpec
+  (composite-generator [s params]
+    (macros/assert! false "You must provide a leaf generator for %s" s)))
 
-;; TODO: do stuff with guards and preconditions?
-(extend-protocol Generator
-  schema.spec.variant.VariantSpec
-  (generator [s params]
-    (generators/one-of
-     (for [o (macros/safe-get s :options)]
-       (sub-generator o params))))
 
-  schema.spec.collection.CollectionSpec
-  (generator [s params]
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Public
+
+(def Schema
+  "A Schema for Schemas"
+  (s/protocol s/Schema))
+
+(def Generator
+  "A test.check generator"
+  s/Any)
+
+(def LeafGenerators
+  "A mapping from schemas to generating functions that should be used."
+  (s/=> (s/maybe Generator) Schema))
+
+(def gen-rational
+  "Simple generator of rational numbers."
+  (generators/one-of
+   [generators/int
     (generators/fmap
-     (comp (:constructor s) (partial apply concat))
-     (apply
-      generators/tuple
-      (for [{:keys [cardinality] :as e} (:elements s)]
-        (let [g (sub-generator e params)]
-          (case cardinality
-            :exactly-one (generators/fmap vector g)
-            :at-most-one (generators/one-of [(generators/return nil) (generators/fmap vector g)])
-            :zero-or-more (generators/vector g))))))))
-
-(defn schema-generator [s params]
-  (let [sp (s/spec s)]
-    (if (instance? schema.spec.leaf.LeafSpec sp)
-      (leaf-generator s)
-      (generator sp params))))
-
-(defn generate [s]
-  (generators/sample
-   (schema-generator
-    s
-    {:subschema-generator schema-generator
-     :cache (java.util.IdentityHashMap.)}) 10))
+     (fn [[a b]] (/ (* a a a a) b))
```

Comment:
oops, good call.  fuck it, we're doing five multiplications http://www.theonion.com/blogpost/fuck-everything-were-doing-five-blades-11056